compiler/rustc_resolve/src/ident.rs

```diff
@@ -1502,10 +1502,10 @@ impl<'a> Resolver<'a> {
                         return PathResult::NonModule(PartialRes::new(Res::Err));
                     } else if opt_ns.is_some() && (is_last || maybe_assoc) {
                         self.lint_if_path_starts_with_module(finalize, path, second_binding);
-                        return PathResult::NonModule(PartialRes::with_unresolved_segments(
-                            res,
-                            path.len() - i - 1,
-                        ));
+                        let partial_res =
+                            PartialRes::with_unresolved_segments(res, path.len() - i - 1);
+                        record_segment_res(self, partial_res);
```

Comment:
```suggestion
                        record_segment_res(self, PartialRes::new(res));
```
For the *segment* it's a full resolution, it's partial only for the whole path.

It should also help to keep the panic in `fn expect_full_res`, which should panic, otherwise it wouldn't be called `expect_*`.